internal/openurl/openurl.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package openurl provides helpers for URLMux and URLOpeners in portable APIs.
+package openurl // import "gocloud.dev/internal/openurl"
+
+import (
+	"fmt"
+	"net/url"
+)
+
+// SchemeMap maps URL schemes to values. The zero value is an empty map.
+type SchemeMap map[string]interface{}
+
+// Register registers scheme for value; subsequent calls to FromString or
+// FromURL with scheme will return value.
+// api is the portable API name (e.g., "blob") and typ is the portable type
+// (e.g., "Bucket").
+// Register panics if scheme has already been registered.
```

Comment:
It just occurred to me that you could store api and typ in the map. That way the zero value would still be good to go, and you wouldn't have to dup api and typ in the FromXXX calls.

Also, you might as well have a single arg that is of the form api.type, since you always use them together. (Or take the two args here, but fmt.Sprintf them together before you store them.)

Reply:
Oh, saw your comment above. So yeah, I would take two args in Register, build and store the right string, and take nothing in the Froms.

Reply:
Do you mean store it in the map at some hardcoded-key?

I have a follow-on PR where I'm going to make it so that there's only a single `URLOpener` interface per portable API ("pubsub"), instead of one per type ("pubsub/Topic", "pubsub/Subscription"). So you'd `Register` an `URLOpener` for "pubsub", not for a specific type, and the `URLOpener` could have multiple functions (`OpenTopic`, `OpenSubscription`). The functionality ends up being the same, but less boilerplate.

I think this means that `Register` could take the "api" and do what you suggest, while the `From` functions would still need to take the "typ".

Sound reasonable? If so, I'd prefer to add a TODO here and do it in the follow-on PR.

Reply:
Sounds reasonable, and I'm fine with a TODO.

Reply:
Done.